### PR TITLE
Windows: Add-Warning() and Add-DeprecationWarning()

### PR DESCRIPTION
**Pull request by dagwieers**
##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

module_utils/powershell.ps1

##### ANSIBLE VERSION

v2.3

##### SUMMARY

Added an Add-Warning() function to assist modules in managing warnings. This functions ensures that if there is not warnings-attribute (or if it isn't a list) the attribute is added correctly and the message is appended to the list.

Similar to what already exists for python modules.

### CODE REVIEW

**Comment by jhawkesworth** [comment]
This is great, but just wondering if we should stick with powershell naming convention i.e. Verb-Noun so instead of Warn you'd probably have something like Write-Warning or perhaps Write-AnsibleWarning?

Personally its not something I'm particularly keen on but worth considering, because as soon as its out there its going to get used (has been already in your other PRs) - its effectively adding to the ansible module (powershell) API.  Its a tricky balance as your function names would be more familiar to python module authors.

**Comment by dagwieers** [comment]
The whole Windows module infrastructure is going to be rewritten by @nitzmahone anyhow, but in this case I wanted to stay close to the python modules. I try to do the same for the option to Set-AnsibleParam for instance, I hate the `--failifempty` (should have been `-required`) or `-ValidateSet` (should have been `-choices`).

So I understand that the Windows/Powershell world has their own way of doing things, but I wouldn't deviate from what Ansible is doing. IMO it is going to be much easier to document this if we stick to common practices as much as possible and doesn't alienate one group from the other.

I would have preferred warn() instead, but that's less an issue because Powershell doesn't care anyhow :-)

That said, I do agree that we want to get this right before we put this out (in a real release). That's a very important concern for me as well.

**Comment by bcoca** [comment]
@nitzmahone is on PTO right now, but I believe he'll agree that we keep the API as close as possible.

**Comment by jhawkesworth** [comment]
Wouldn't want to hold up this going in, just wanted to flag up concern.  More important to have the api capability than to make it instantly recognisable to those familiar with powershell.

shipit

**Comment by dagwieers** [comment]
@jhawkesworth No worries, we can wait on @nitzmahone 's return to decide. I know some of the interfaces will be changed in the future, but since we don't know whether this will hit v2.3 or later I prefer we don't focus too much on that right now and add as much functionality to make it on-par with Python modules.

It is going to be much easier to replace A with B once it is needed, then to have to revisit the code and refactor such support into it after the fact.

**Comment by nitzmahone** [comment]
My only issue with adding new global functions to to the API surface area at this point is that anything we add will have to get dragged around for at least two releases before we could remove it (since we can't know what private modules are using). OTOH, since the new wrapper supports arbitrary module_utils inclusions (and powershell.ps1 is not strictly required anymore), anything new won't be defined in powershell.ps1 anyway, so we can pretty much consider that file "legacy" already. 

As far as naming goes, I think @jhawkesworth is on the right track- I'd vote for `Add-Warning`/`Add-DeprecationWarning` (since we're not "writing" it per-se like Write-Output/Error/etc). 

I'm OK to merge this with that name change (or feel free to suggest others)...

**Comment by dagwieers** [comment]
@nitzmahone Thanks for the feedback. Implemented as requested.

**Comment by jhawkesworth** [comment]
Nice.

shipit

File: lib/ansible/modules/windows/win_regedit.ps1 [hunk 1]
```diff
@@ -37,12 +37,11 @@ $result = @{
     diff = @{
         prepared = ""
     }
-    warnings = @()
 }
 
 # Fix HCCC:\ PSDrive for pre-2.3 compatibility
 if ($path -match "^HCCC:\\") {
-    $result.warnings += "Please use path: HKCC:\... instead of path: $path\n"
+    Warn $result "Please use path: HKCC:\... instead of path: $path"
```
Comment (nitzmahone) [inline lib/ansible/modules/windows/win_regedit.ps1]: Needs rename :)

File: lib/ansible/modules/windows/win_command.ps1 [hunk 2]
```diff
@@ -105,6 +103,11 @@ $util_type = Add-Type -TypeDefinition $util_def
 
 # FUTURE: extract this code to separate module_utils as Windows module API version of run_command
 
+$result = @{
```
Comment (nitzmahone) [inline lib/ansible/modules/windows/win_command.ps1]: I don't really have a problem with this placement, but this seems like an arbitrary and completely unrelated change.

File: lib/ansible/modules/windows/win_shell.ps1 [hunk 3]
```diff
@@ -80,6 +78,11 @@ If($removes -and -not $(Test-Path $removes)) {
 
 Add-Type -TypeDefinition $helper_def
 
+$result = @{
```
Comment (nitzmahone) [inline lib/ansible/modules/windows/win_shell.ps1]: Ditto

**Review by nitzmahone — CHANGES_REQUESTED** [review]

**Review by nitzmahone — APPROVED** [review]

File: lib/ansible/modules/windows/win_regedit.ps1 (hunk 1 above)
Comment (dagwieers) [reply lib/ansible/modules/windows/win_regedit.ps1]: I didn't realize these changes were in this PR too.

File: lib/ansible/modules/windows/win_command.ps1 (hunk 2 above)
Comment (dagwieers) [reply lib/ansible/modules/windows/win_command.ps1]: I didn't realize these changes were in this PR too.

File: lib/ansible/modules/windows/win_shell.ps1 (hunk 3 above)
Comment (dagwieers) [reply lib/ansible/modules/windows/win_shell.ps1]: I didn't realize these changes were in this PR too.

**Pull request merged**